File: solutions/cpp/KthLargestElementInAnArray.cc
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        int left, right;

        left = 0;
        right = nums.size() - 1;
        int pivot = partition(nums, left, right);

        while (pivot != k-1) {
            if (pivot < k-1) {
                left = pivot + 1;
                pivot = partition(nums, left, right);
            } else {
                right = pivot - 1;
                pivot = partition(nums, left, right);
            }
        }

        return nums[pivot];
    }

private:
    int partition(vector<int> &arr, int left, int right) {
        int last = left;

        std::swap(arr[left], arr[(left+right)/2]);

        for (int i = left+1; i <= right; ++i) {
            if (arr[i] > arr[left]) {
                std::swap(arr[i], arr[++last]);
            }
        }

        std::swap(arr[left], arr[last]);

        return last;
    }
};

File: solutions/go/KthLargestElementInAnArray.go
func findKthLargest(nums []int, k int) int {
	pivot := partition(nums)

	if pivot < k-1 {
		return findKthLargest(nums[pivot+1:], k-pivot-1)
	} else if pivot > k-1 {
		return findKthLargest(nums[:pivot], k)
	} else {
		return nums[pivot]
	}
}

func partition(arr []int) int {
	length := len(arr)

	last := 0
	swap(arr, 0, length/2)

	for i := 1; i < length; i++ {
		if arr[i] > arr[0] {
			last++
			swap(arr, i, last)
		}
	}

	swap(arr, 0, last)

	return last
}

func swap(arr []int, i, j int) {
	arr[i], arr[j] = arr[j], arr[i]
}
